File: driver-core/src/test/unit/com/mongodb/client/ImmutableDocumentCodec.java
/*
 * Copyright 2016 MongoDB, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mongodb.client;

import org.bson.BsonReader;
import org.bson.BsonValue;
import org.bson.BsonWriter;
import org.bson.Document;
import org.bson.codecs.CollectibleCodec;
import org.bson.codecs.DecoderContext;
import org.bson.codecs.EncoderContext;
import org.bson.codecs.configuration.CodecRegistry;
import org.bson.types.ObjectId;

import java.util.LinkedHashMap;

import static java.lang.String.format;

public final class ImmutableDocumentCodec implements CollectibleCodec<ImmutableDocument> {
    private final CodecRegistry codecRegistry;
    private static final String ID_FIELD_NAME = "_id";

    public ImmutableDocumentCodec(final CodecRegistry codecRegistry) {
        this.codecRegistry = codecRegistry;
    }

    @Override
    public ImmutableDocument generateIdIfAbsentFromDocument(final ImmutableDocument document) {
        LinkedHashMap<String, Object> mutable = new LinkedHashMap<String, Object>(document);
        mutable.put(ID_FIELD_NAME, new ObjectId());
        return new ImmutableDocument(mutable);
    }

    @Override
    public boolean documentHasId(final ImmutableDocument document) {
        return document.containsKey(ID_FIELD_NAME);
    }

    @Override
    public BsonValue getDocumentId(final ImmutableDocument document) {
        if (!documentHasId(document)) {
            throw new IllegalStateException(format("The document does not contain an %s", ID_FIELD_NAME));
        }
        return document.toBsonDocument(ImmutableDocument.class, codecRegistry).get(ID_FIELD_NAME);
    }

    @Override
    public void encode(final BsonWriter writer, final ImmutableDocument value, final EncoderContext encoderContext) {
        codecRegistry.get(Document.class).encode(writer, new Document(value), encoderContext);
    }

    @Override
    public Class<ImmutableDocument> getEncoderClass() {
        return ImmutableDocument.class;
    }

    @Override
    public ImmutableDocument decode(final BsonReader reader, final DecoderContext decoderContext) {
        Document document = codecRegistry.get(Document.class).decode(reader, decoderContext);
        return new ImmutableDocument(document);
    }
}

File: bson-scala/src/main/scala/org/mongodb/scala/bson/codecs/ImmutableDocumentCodec.scala
/*
 * Copyright 2008-present MongoDB, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.mongodb.scala.bson.codecs

import org.bson.codecs.configuration.CodecRegistry
import org.bson.codecs.{ BsonDocumentCodec, CollectibleCodec, DecoderContext, EncoderContext }
import org.bson.{ BsonReader, BsonValue, BsonWriter }
import org.mongodb.scala.bson.collection.immutable.Document

/**
 * Companion helper for immutable Document instances.
 */
object ImmutableDocumentCodec {
  def apply(): ImmutableDocumentCodec = ImmutableDocumentCodec(None)
  def apply(registry: CodecRegistry): ImmutableDocumentCodec = ImmutableDocumentCodec(Some(registry))
}

/**
 * A Codec for immutable Document instances.
 *
 * As the underlying driver expects documents to be mutable the driver has direct access to the Documents underlying
 * mutable `BsonDocument` instance and therefore will mutate the document when adding an `_id`
 */
case class ImmutableDocumentCodec(registry: Option[CodecRegistry]) extends CollectibleCodec[Document] {

  lazy val underlying: BsonDocumentCodec = {
    registry.map(new BsonDocumentCodec(_)).getOrElse(new BsonDocumentCodec)
  }

  override def generateIdIfAbsentFromDocument(document: Document): Document = {
    if (!underlying.documentHasId(document.underlying)) {
      Document(underlying.generateIdIfAbsentFromDocument(document.toBsonDocument.clone))
    } else {
      document
    }
  }

  override def documentHasId(document: Document): Boolean = underlying.documentHasId(document.underlying)

  override def getDocumentId(document: Document): BsonValue = underlying.getDocumentId(document.underlying)

  override def encode(writer: BsonWriter, value: Document, encoderContext: EncoderContext): Unit =
    underlying.encode(writer, value.underlying, encoderContext)

  override def getEncoderClass: Class[Document] = classOf[Document]

  override def decode(reader: BsonReader, decoderContext: DecoderContext): Document =
    Document(underlying.decode(reader, decoderContext))
}
